include/zephyr/kernel/obj_core.h

```diff
@@ -0,0 +1,283 @@
+#ifndef __KERNEL_OBJ_CORE_H__
+#define __KERNEL_OBJ_CORE_H__
+
+#include <zephyr/sys/dlist.h>
+
+/**
+ * @brief Macro to convert kernel object into pointer to object core
+ *
+ * This macro is used to convert a pointer to a kernel object (e.g. thread or
+ * memory block) into a pointer to its object core.
+ */
+#define K_OBJ_CORE(kobj)  (&((kobj)->obj_core))
+
+/* Known kernel object types */
+
+#define OBJ_TYPE_CONDVAR_ID    0x434f4e44   /* "COND" in ASCII */
+#define OBJ_TYPE_CPU_ID        0x4350555f   /* "CPU_" in ASCII */
+#define OBJ_TYPE_EVENT_ID      0x45564e54   /* "EVNT" in ASCII */
+#define OBJ_TYPE_FIFO_ID       0x4649464f   /* "FIFO" in ASCII */
+#define OBJ_TYPE_KERNEL_ID     0x4B524e4c   /* "KRNL" in ASCII */
+#define OBJ_TYPE_LIFO_ID       0x4c49464f   /* "LIFO" in ASCII */
+#define OBJ_TYPE_MEM_BLOCK_ID  0x4d424c4b   /* "MBLK" in ASCII */
+#define OBJ_TYPE_MBOX_ID       0x4d424f58   /* "MBOX" in ASCII */
+#define OBJ_TYPE_MEM_SLAB_ID   0x534c4142   /* "SLAB" in ASCII */
+#define OBJ_TYPE_MSGQ_ID       0x4d534751   /* "MSGQ" in ASCII */
+#define OBJ_TYPE_MUTEX_ID      0x4d555458   /* "MUTX" in ASCII */
+#define OBJ_TYPE_PIPE_ID       0x50495045   /* "PIPE" in ASCII */
+#define OBJ_TYPE_SEM_ID        0x53454d34   /* "SEM4" in ASCII */
+#define OBJ_TYPE_STACK_ID      0x5354434b   /* "STCK" in ASCII */
+#define OBJ_TYPE_THREAD_ID     0x54485244   /* "THRD" in ASCII */
+#define OBJ_TYPE_TIMER_ID      0x54494d52   /* "TIMR" in ASCII */
```

Comment:
This is cute and all, but if you're going to go through the trouble, I demand that this computation be performatively elegant and not manual:

```c
#define KOBJ_TYPE_ID_NONSENSE(s) (s[0]<<24)|(s[1]<<16)|(s[2]<<8)|(s[3])
#define OBJ_TYPE_CONDVAR_ID KOBJ_TYPE_ID_NONSENSE("COND")
```

Also, I note that these values are all expressed... big-endian?!

Reply:
Seriously but pedantically: "OBJ_" probably isn't a unique enough prefix and we need a K_ on there too (or just a double underscore if I'm understanding the current direction of our style guide?).  This is a /include header that can find its way into user code, so everything needs to be namespaced, even "internal" values.

Reply:
Good points. Will update.